### PR TITLE
Add landmarks to the state and deserialize them.

### PR DESCRIPTION
**Pull request by pifon2a**
Also add a boolean indicator whether to optimize landmarks or not.

### CODE REVIEW

File: cartographer/mapping/internal/2d/pose_graph_2d.cc [hunk 1]
```diff
@@ -418,6 +418,11 @@ void PoseGraph2D::FreezeTrajectory(const int trajectory_id) {
   });
 }
 
+void PoseGraph2D::FreezeLandmarks() {
+  common::MutexLocker locker(&mutex_);
+  freeze_landmarks_ = true;
```
Comment (ojura) [inline cartographer/mapping/internal/2d/pose_graph_2d.cc]: Do not modify freeze_landmarks outside of worker queue.

File: cartographer/mapping/internal/2d/pose_graph_2d.cc [hunk 2]
```diff
@@ -554,7 +559,7 @@ void PoseGraph2D::RunOptimization() {
   // is time consuming, so not taking the mutex before Solve to avoid blocking
   // foreground processing.
   optimization_problem_.Solve(constraints_, frozen_trajectories_,
-                              landmark_nodes_);
+                              landmark_nodes_, freeze_landmarks_);
```
Comment (ojura) [inline cartographer/mapping/internal/2d/pose_graph_2d.cc]: Update comment above, add freeze_landmarks to the list of variables that we promise no other thread is accessing.

**Comment by pifon2a** [comment]
@ojura done. Thanks!

File: cartographer/mapping/internal/2d/pose_graph_2d.h [hunk 3]
```diff
@@ -257,6 +258,9 @@ class PoseGraph2D : public PoseGraph {
   // Set of all frozen trajectories not being optimized.
   std::set<int> frozen_trajectories_ GUARDED_BY(mutex_);
 
+  // Whether or not optimize landmark poses.
+  bool freeze_landmarks_ GUARDED_BY(mutex_) = false;
```
Comment (ojura) [inline cartographer/mapping/internal/2d/pose_graph_2d.h]: IIRC default values are to be avoided, according to the Google coding style.

File: cartographer/mapping/internal/2d/pose_graph/optimization_problem_2d.cc [hunk 4]
```diff
@@ -232,8 +244,8 @@ void OptimizationProblem2D::Solve(
         C_nodes.at(constraint.node_id).data());
   }
   // Add cost  functions for landmarks.
```
Comment (ojura) [inline cartographer/mapping/internal/2d/pose_graph/optimization_problem_2d.cc]: Nit: remove space after cost.

File: cartographer/mapping/internal/3d/pose_graph/optimization_problem_3d.h [hunk 5]
```diff
@@ -92,7 +92,8 @@ class OptimizationProblem3D {
   // Optimizes the global poses.
   void Solve(const std::vector<Constraint>& constraints,
              const std::set<int>& frozen_trajectories,
-             const std::map<std::string, LandmarkNode>& landmark_nodes);
+             const std::map<std::string, LandmarkNode>& landmark_nodes,
+             bool freeze_landmarks);
```
Comment (cschuet) [inline cartographer/mapping/internal/3d/pose_graph/optimization_problem_3d.h]: As discussed you can infer the that of `freeze_landmarks` from the state of `frozen_trajectories`.

File: cartographer/mapping/internal/2d/pose_graph_2d.h (hunk 3 above)
Comment (pifon2a) [reply cartographer/mapping/internal/2d/pose_graph_2d.h]: Style guide talks mostly about [default arguments](https://google.github.io/styleguide/cppguide.html#Default_Arguments) in functions. They are allowed now (it was changed a 1+ year ago).

**Comment by ojura** [comment]
lgtm. I could imagine someone perhaps wanting to add additional landmarks on top of a frozen trajectory, but that can be added later.

**Pull request merged**